File: flutter/packages/simple_brilliant_app/example/imu_compass/test/mag_calibration_test.dart
import 'dart:convert';
import 'dart:math';

import 'package:flutter_test/flutter_test.dart';
import 'package:imu_compass/compass_heading.dart';
import 'package:imu_compass/mag_calibration.dart';

void main() {
  group('default alignment', () {
    test('a fresh calibration ships the Halo default matrix (not identity)', () {
      final cal = MagCalibration();
      expect(cal.matrix, MagCalibration.haloDefaultAlignment);
      expect(cal.hasAlignment, isTrue);
    });

    test('the default alignment is a valid rotation (orthonormal, det +1)', () {
      const m = MagCalibration.haloDefaultAlignment;
      for (var i = 0; i < 3; i++) {
        for (var j = 0; j < 3; j++) {
          final dot = m[i][0] * m[j][0] + m[i][1] * m[j][1] + m[i][2] * m[j][2];
          expect(dot, closeTo(i == j ? 1.0 : 0.0, 1e-4));
        }
      }
      final det = m[0][0] * (m[1][1] * m[2][2] - m[1][2] * m[2][1]) -
          m[0][1] * (m[1][0] * m[2][2] - m[1][2] * m[2][0]) +
          m[0][2] * (m[1][0] * m[2][1] - m[1][1] * m[2][0]);
      expect(det, closeTo(1.0, 1e-4));
    });
  });

  group('MagCalibration.apply', () {
    test('identity + zero offset is a no-op', () {
      // the class default is now the Halo alignment, so pass identity explicitly
      final cal = MagCalibration(matrix: MagCalibration.identity);
      expect(cal.hasAlignment, isFalse);
      final (x, y, z) = cal.apply(10, -20, 30);
      expect(x, closeTo(10, 1e-9));
      expect(y, closeTo(-20, 1e-9));
      expect(z, closeTo(30, 1e-9));
    });

    test('hard-iron offset is added before the matrix', () {
      final cal = MagCalibration(offset: [1, 2, 3], matrix: MagCalibration.identity);
      final (x, y, z) = cal.apply(10, 20, 30);
      expect(x, closeTo(11, 1e-9));
      expect(y, closeTo(22, 1e-9));
      expect(z, closeTo(33, 1e-9));
    });

    test('matrix @ (m + offset): a 90° z-rotation maps +X→+Y', () {
      // rotation of +90° about z: (x,y,z) -> (-y, x, z)
      final cal = MagCalibration(matrix: const [
        [0, -1, 0],
        [1, 0, 0],
        [0, 0, 1],
      ]);
      expect(cal.hasAlignment, isTrue);
      final (x, y, z) = cal.apply(1, 0, 0);
      expect(x, closeTo(0, 1e-9));
      expect(y, closeTo(1, 1e-9));
      expect(z, closeTo(0, 1e-9));
    });
  });

  group('MagCalibration JSON', () {
    test('round-trips through JSON with the Python tool key names', () {
      final cal = MagCalibration(
        offset: [1.5, -2.5, 0.25],
        matrix: const [
          [0, -1, 0],
          [1, 0, 0],
          [0, 0, 1],
        ],
        headingOffsetDeg: 42.0,
        fieldUT: 960.0,
      );
      final restored = MagCalibration.fromJsonString(cal.toJsonString());
      expect(restored.offset, cal.offset);
      expect(restored.matrix, cal.matrix);
      expect(restored.headingOffsetDeg, cal.headingOffsetDeg);
      expect(restored.fieldUT, cal.fieldUT);
    });

    test('parses the exact key names emitted by calibration.py', () {
      const json = '{"offset":[1,2,3],'
          '"matrix":[[1,0,0],[0,1,0],[0,0,1]],'
          '"heading_offset_deg":12.7,"field_uT":57.0}';
      final cal = MagCalibration.fromJson(jsonDecode(json) as Map<String, dynamic>);
      expect(cal.offset, [1.0, 2.0, 3.0]);
      expect(cal.headingOffsetDeg, 12.7);
      expect(cal.fieldUT, 57.0);
    });

    test('tolerates missing heading_offset_deg / field_uT', () {
      const json = '{"offset":[0,0,0],"matrix":[[1,0,0],[0,1,0],[0,0,1]]}';
      final cal = MagCalibration.fromJson(jsonDecode(json) as Map<String, dynamic>);
      expect(cal.headingOffsetDeg, 0.0);
      expect(cal.fieldUT, 0.0);
    });
  });

  group('MagCalibration.heading', () {
    test('heading offset shifts the reported heading', () {
      // level device (+Z up), field pointing along +Y (forward) => 0°.
      const mag = [0.0, 30.0, 0.0];
      const accel = [0.0, 0.0, 1.0];
      final base = MagCalibration(matrix: MagCalibration.identity);
      expect(base.heading(mag[0], mag[1], mag[2], accel[0], accel[1], accel[2]),
          closeTo(0.0, 1e-6));

      final shifted = base.copyWith(headingOffsetDeg: 90.0);
      expect(
          shifted.heading(mag[0], mag[1], mag[2], accel[0], accel[1], accel[2]),
          closeTo(90.0, 1e-6));
    });

    test('declination is applied on top of the heading offset', () {
      const mag = [0.0, 30.0, 0.0];
      const accel = [0.0, 0.0, 1.0];
      final cal = MagCalibration(headingOffsetDeg: 10.0, matrix: MagCalibration.identity);
      expect(
          cal.heading(mag[0], mag[1], mag[2], accel[0], accel[1], accel[2],
              declination: 12.7),
          closeTo(22.7, 1e-6));
    });
  });

  group('set-North offset math (mirrors imu_heading.py --set-north)', () {
    // The app measures the live heading (which already includes the current
    // offset) and picks newOffset = currentOffset - measured so the pose reads 0.
    double reZero(MagCalibration cal, double mx, double my, double mz,
        double ax, double ay, double az) {
      final measured = cal.heading(mx, my, mz, ax, ay, az);
      return (cal.headingOffsetDeg - measured) % 360.0;
    }

    test('re-zeros an arbitrary pose to read ~0° afterwards', () {
      // device facing East-ish: field along +X => uncalibrated heading 90°.
      const mag = [30.0, 0.0, 0.0];
      const accel = [0.0, 0.0, 1.0];
      var cal = MagCalibration();
      final newOffset = reZero(cal, mag[0], mag[1], mag[2], accel[0], accel[1], accel[2]);
      cal = cal.copyWith(headingOffsetDeg: newOffset);
      final after = cal.heading(mag[0], mag[1], mag[2], accel[0], accel[1], accel[2]);
      expect(min(after, 360 - after), closeTo(0.0, 1e-6));
    });

    test('re-zero is stable when an offset is already present', () {
      const mag = [30.0, 0.0, 0.0];
      const accel = [0.0, 0.0, 1.0];
      var cal = MagCalibration(headingOffsetDeg: 137.0);
      final newOffset = reZero(cal, mag[0], mag[1], mag[2], accel[0], accel[1], accel[2]);
      cal = cal.copyWith(headingOffsetDeg: newOffset);
      final after = cal.heading(mag[0], mag[1], mag[2], accel[0], accel[1], accel[2]);
      expect(min(after, 360 - after), closeTo(0.0, 1e-6));
    });
  });

  group('gravity normalization in tilt compensation', () {
    test('raw accel (magnitude ~960) gives the same heading as a unit vector', () {
      const mag = [12.0, 34.0, -5.0];
      final unit = CompassHeading.calculateTiltCompensatedHeading(
        magX: mag[0], magY: mag[1], magZ: mag[2],
        gravityX: 0.0, gravityY: 0.0, gravityZ: 1.0);
      final raw = CompassHeading.calculateTiltCompensatedHeading(
        magX: mag[0], magY: mag[1], magZ: mag[2],
        gravityX: 0.0, gravityY: 0.0, gravityZ: 960.0);
      expect(raw, closeTo(unit, 1e-6));
    });
  });
}

File: webbluetooth/packages/brilliant-msg/tests/mag-calibration.test.mjs
// Unit tests for MagCalibration + CompassHeading. No device or test framework
// required: run with `npm test` (node --test) — the pretest script builds the
// dist bundle these tests import (and the linked brilliant-ble, if needed).
// Mirrors the Flutter example's test/mag_calibration_test.dart and the Python
// tests/test_calibration.py guards so all three SDKs stay in lockstep.
import { test } from 'node:test';
import assert from 'node:assert/strict';

// The published bundle targets the browser and references `self` at module-eval
// time (image-js). MagCalibration/CompassHeading have no browser deps, so shim
// `self` and dynamic-import after it is set (static imports are hoisted).
globalThis.self ??= globalThis;
const { MagCalibration, CompassHeading } = await import('../dist/brilliant-msg.es.js');

const close = (actual, expected, eps = 1e-6) =>
    assert.ok(Math.abs(actual - expected) <= eps,
        `expected ${actual} to be within ${eps} of ${expected}`);

test('default alignment: a fresh calibration ships the Halo default matrix (not identity)', () => {
    const cal = new MagCalibration();
    assert.deepEqual(cal.matrix, MagCalibration.haloDefaultAlignment);
    assert.equal(cal.hasAlignment, true);
});

test('default alignment is a valid rotation (orthonormal, det +1)', () => {
    const m = MagCalibration.haloDefaultAlignment;
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            const dot = m[i][0] * m[j][0] + m[i][1] * m[j][1] + m[i][2] * m[j][2];
            close(dot, i === j ? 1.0 : 0.0, 1e-4);
        }
    }
    const det = m[0][0] * (m[1][1] * m[2][2] - m[1][2] * m[2][1]) -
        m[0][1] * (m[1][0] * m[2][2] - m[1][2] * m[2][0]) +
        m[0][2] * (m[1][0] * m[2][1] - m[1][1] * m[2][0]);
    close(det, 1.0, 1e-4);
});

test('apply: identity + zero offset is a no-op', () => {
    const cal = new MagCalibration({ matrix: MagCalibration.identity });
    assert.equal(cal.hasAlignment, false);
    const [x, y, z] = cal.apply(10, -20, 30);
    close(x, 10);
    close(y, -20);
    close(z, 30);
});

test('apply: hard-iron offset is added before the matrix', () => {
    const cal = new MagCalibration({ offset: [1, 2, 3], matrix: MagCalibration.identity });
    const [x, y, z] = cal.apply(10, 20, 30);
    close(x, 11);
    close(y, 22);
    close(z, 33);
});

test('apply: matrix @ (m + offset): a 90deg z-rotation maps +X->+Y', () => {
    // rotation of +90deg about z: (x,y,z) -> (-y, x, z)
    const cal = new MagCalibration({ matrix: [[0, -1, 0], [1, 0, 0], [0, 0, 1]] });
    assert.equal(cal.hasAlignment, true);
    const [x, y, z] = cal.apply(1, 0, 0);
    close(x, 0);
    close(y, 1);
    close(z, 0);
});

test('JSON: round-trips with the Python tool key names', () => {
    const cal = new MagCalibration({
        offset: [1.5, -2.5, 0.25],
        matrix: [[0, -1, 0], [1, 0, 0], [0, 0, 1]],
        headingOffsetDeg: 42.0,
        fieldUT: 960.0,
    });
    const restored = MagCalibration.fromJsonString(cal.toJsonString());
    assert.deepEqual(restored.offset, cal.offset);
    assert.deepEqual(restored.matrix, cal.matrix);
    assert.equal(restored.headingOffsetDeg, cal.headingOffsetDeg);
    assert.equal(restored.fieldUT, cal.fieldUT);
});

test('JSON: parses the exact key names emitted by calibration.py', () => {
    const json = '{"offset":[1,2,3],'
        + '"matrix":[[1,0,0],[0,1,0],[0,0,1]],'
        + '"heading_offset_deg":12.7,"field_uT":57.0}';
    const cal = MagCalibration.fromJsonString(json);
    assert.deepEqual(cal.offset, [1.0, 2.0, 3.0]);
    assert.equal(cal.headingOffsetDeg, 12.7);
    assert.equal(cal.fieldUT, 57.0);
});

test('JSON: tolerates missing heading_offset_deg / field_uT', () => {
    const json = '{"offset":[0,0,0],"matrix":[[1,0,0],[0,1,0],[0,0,1]]}';
    const cal = MagCalibration.fromJsonString(json);
    assert.equal(cal.headingOffsetDeg, 0.0);
    assert.equal(cal.fieldUT, 0.0);
});

test('heading: heading offset shifts the reported heading', () => {
    // level device (+Z up), field pointing along +Y (forward) => 0deg.
    const mag = [0.0, 30.0, 0.0];
    const accel = [0.0, 0.0, 1.0];
    const base = new MagCalibration({ matrix: MagCalibration.identity });
    close(base.heading(mag[0], mag[1], mag[2], accel[0], accel[1], accel[2]), 0.0);

    const shifted = base.copyWith({ headingOffsetDeg: 90.0 });
    close(shifted.heading(mag[0], mag[1], mag[2], accel[0], accel[1], accel[2]), 90.0);
});

test('heading: declination is applied on top of the heading offset', () => {
    const mag = [0.0, 30.0, 0.0];
    const accel = [0.0, 0.0, 1.0];
    const cal = new MagCalibration({ headingOffsetDeg: 10.0, matrix: MagCalibration.identity });
    close(cal.heading(mag[0], mag[1], mag[2], accel[0], accel[1], accel[2], 12.7), 22.7);
});

test('set-North: re-zeros an arbitrary pose to read ~0deg afterwards', () => {
    // The app measures the live heading (which already includes the current
    // offset) and picks newOffset = currentOffset - measured so the pose reads 0.
    const reZero = (cal, mx, my, mz, ax, ay, az) => {
        const measured = cal.heading(mx, my, mz, ax, ay, az);
        return ((cal.headingOffsetDeg - measured) % 360.0 + 360.0) % 360.0;
    };
    // device facing East-ish: field along +X => uncalibrated heading 90deg.
    const mag = [30.0, 0.0, 0.0];
    const accel = [0.0, 0.0, 1.0];
    let cal = new MagCalibration({ headingOffsetDeg: 137.0, matrix: MagCalibration.identity });
    const newOffset = reZero(cal, mag[0], mag[1], mag[2], accel[0], accel[1], accel[2]);
    cal = cal.copyWith({ headingOffsetDeg: newOffset });
    const after = cal.heading(mag[0], mag[1], mag[2], accel[0], accel[1], accel[2]);
    close(Math.min(after, 360 - after), 0.0);
});

test('gravity normalization: raw accel (~960) gives the same heading as a unit vector', () => {
    const mag = [12.0, 34.0, -5.0];
    const unit = CompassHeading.calculateTiltCompensatedHeading(
        mag[0], mag[1], mag[2], 0.0, 0.0, 1.0);
    const raw = CompassHeading.calculateTiltCompensatedHeading(
        mag[0], mag[1], mag[2], 0.0, 0.0, 960.0);
    close(raw, unit);
});
